src/main/java/com/android/volley/CacheDispatcher.java

```diff
@@ -154,4 +180,92 @@ public void run() {
+        if (response.cacheEntry == null || response.cacheEntry.isExpired()) {
+            onNoUsableResponseReceived(request);
+            return;
+        }
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+        }
+        if (waitingRequests != null) {
+            if (VolleyLog.DEBUG) {
+                VolleyLog.v("Releasing %d waiting requests for cacheKey=%s.",
+                    waitingRequests.size(), cacheKey);
+            }
+            // Process all queued up requests.
+            for (Request<?> waiting : waitingRequests) {
+                mDelivery.postResponse(waiting, response);
+            }
+        }
+    }
+
+    /** No valid response received from network, release waiting requests. */
+    @Override
+    public void onNoUsableResponseReceived(Request<?> request) {
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+            if (waitingRequests != null) {
+                if (VolleyLog.DEBUG) {
+                    VolleyLog.v("%d waiting requests for cacheKey=%s; resend to network",
+                        waitingRequests.size(), cacheKey);
+                }
+                Request<?> nextInLine = waitingRequests.remove();
+                mWaitingRequests.put(cacheKey, waitingRequests);
+                try {
+                    mNetworkQueue.put(nextInLine);
+                } catch (InterruptedException iex) {
+                    VolleyLog.e("Couldn't add request to queue. %s", iex.toString());
+                    // Restore the interrupted status
+                    Thread.currentThread().interrupt();
+                    quit();
+                }
+            }
+        }
+    }
+
+    /**
+     * For cacheable requests, if a request for the same cache key is already in flight,
+     * add it to a queue to wait for that in-flight request to finish.
+     * @return whether the request was queued. If false, we should continue issuing the request
+     * over the network. If true, we should put the request on hold to be processed when
+     * the in-flight request finishes.
+     */
+    private boolean maybeAddToWaitingRequests(Request<?> request) {
+        String cacheKey = request.getCacheKey();
+        synchronized (mWaitingRequests) {
+            // Insert request into stage if there's already a request with the same cache key
+            // in flight.
+            if (mWaitingRequests.containsKey(cacheKey)) {
+                // There is already a request in flight. Queue up.
+                Queue<Request<?>> stagedRequests = mWaitingRequests.get(cacheKey);
+                if (stagedRequests == null) {
+                    stagedRequests = new LinkedList<Request<?>>();
```

Comment:
ArrayList is more performant here (and virtually everywhere). In other words, LinkedList is best avoided in Java.

Reply:
This one was also just copied from the old code... but in this case I'm fine with making the change as I do think it's reasonable enough.

Reply:
This is a Queue though. Should I make the Queue a List instead?

Reply:
@uhager Ah. Use `ArrayDeque`:

```
This class is likely to be faster than Stack when used as a stack,
and faster than LinkedList when used as a queue.
```

Null items are prohibited. I hope that's not a problem.

Reply:
Volley's minSdkVersion is still 8, and ArrayDeque was added in API 9. Definitely a reasonable change to consider bumping minSdkVersion and making this change but it'd be out of the scope of the CL. So probably best to leave as is for now.